osbuild/loop.py

```diff
@@ -598,21 +681,18 @@ def loop_for_fd(self,
                     continue
 
             try:
-                lo.set_fd(fd)
+                lo.loop_configure(fd, **kwargs)
+            except BlockingIOError:
+                lo.clear_fd()
+                lo.close()
+                continue
             except OSError as e:
                 lo.close()
+                # `loop_configure` returns EBUSY when the pages from the
+                # previously bound file have not been fully cleared yet.
```

Comment:
I think that it would be cleaner (or more obvious) to handle the fallback in case `loop_configure()` raises `OSError` with `errno.ENOTTY` here.

Also the comment will be less ambiguous in such case, because the `EBUSY` can be raised (AFAICT) only if the `loop_configure` falls back to using the "original approach".

Reply:
I think `LOOP_CONFIGURE` would also raise `EBUSY` as the doc says it has the same behavior as `set_status`

Reply:
I thought about it, and since there's two places where this bit of code is done, I think it's still better to keep it in the `Loop` class. My main point is to avoid duplicating some core logic.

However, I've split and renamed the `loop_configure` method. It's now called `set_fd_and_status`, docstring being explicit that the method will try to do the operation atomically if it can and fallback if it can't.

The ioctl call to `LOOP_CONFIGURE` is done within the `configure` method. That way we can unit test this ioctl call without having the fallback involved.

Reply:
> I thought about it, and since there's two places where this bit of code is done, I think it's still better to keep it in the `Loop` class. My main point is to avoid duplicating some core logic.

The logic of handling loop devices is also duplicated in `loop.py` and `remoteloop.py`. I'm always for deduplication of code, but not as the main goal all of the time. Sometimes it can just complicate the code by making it too smart and handling too many cases. I prefer readability and ease of understanding over deduplication where it makes sense. It would be like 4 additional lines, so I do not see it as an issue here.

If other's will approve this as is, I won't object.

> However, I've split and renamed the `loop_configure` method. It's now called `set_fd_and_status`, docstring being explicit that the method will try to do the operation atomically if it can and fallback if it can't.

I'm not sure if I prefer this TBH. The linked LWN page with the patchset says among other things `Renamed LOOP_SET_FD_AND_STATUS to LOOP_CONFIGURE, since the ioctl can now do things LOOP_SET_STATUS couldn't do`

I would still like to see a single purpose method that does only `LOOP_CONFIGURE` ioctl and nothing else. Then use it anywhere you need to. The benefit is that we may end up using it for other purposes in the future if we need to. If you make it too use-case specific and too smart, then this will be harder and we may and up having to refactor it. Personally, I would prefer keeping it as a well defined single-purpose building block that we can then use when handling loopback devices.

Reply:
I missed that comment yesterday. Fair points.